validation_code/ConditionalLuminosityFunction.py

```diff
@@ -63,24 +77,33 @@ def run_validation_test(self, galaxy_catalog, catalog_name, base_output_dir=None
         if prepared is None:
             return TestResult(skipped=True)
 
-        absolute_magnitude_field, quantities_needed = prepared
-        colnames = [absolute_magnitude_field, 'halo_mass', 'redshift_true']
+        absolute_magnitude1_field, absolute_magnitude2_field, quantities_needed = prepared
+        colnames = [absolute_magnitude2_field, 'halo_mass', 'redshift_true']
         bins = (self.magnitude_bins, self.mass_bins, self.z_bins)
         hist_cen = np.zeros((self.n_magnitude_bins, self.n_mass_bins, self.n_z_bins))
         hist_sat = np.zeros_like(hist_cen)
-
+        
+        red_query = GCRQuery((self.color_cut, 
+                              absolute_magnitude1_field, 
+                              absolute_magnitude2_field, 'redshift_true'))
+            
         cen_query = GCRQuery('is_central')
         sat_query = ~cen_query
+
         if 'r_host' in quantities_needed and 'r_vir' in quantities_needed:
             sat_query &= GCRQuery('r_host < r_vir')
 
+        data = next(galaxy_catalog.get_quantities(quantities_needed, return_iterator=True))
+        
         for data in galaxy_catalog.get_quantities(quantities_needed, return_iterator=True):
-            cen_mask = cen_query.mask(data)
+            red_mask = red_query.mask(data)
+            cen_mask = cen_query.mask(data) 
             sat_mask = sat_query.mask(data)
-            data = np.stack((data[k] for k in colnames)).T
-            hist_cen += np.histogramdd(data[cen_mask], bins)[0]
-            hist_sat += np.histogramdd(data[sat_mask], bins)[0]
 
+            data = np.stack((data[k] for k in colnames)).T
+            hist_cen += np.histogramdd(data[cen_mask&red_mask], bins)[0]
+            hist_sat += np.histogramdd(data[sat_mask&red_mask], bins)[0]
```

Comment:
Since red_mask is always applied, you can move then outside the for loop. That is:

At Line 97, write
```python
cen_query &= red_query
sat_query &= red_query
```
Remove Line 99 (`red_mask = red_query.mask(data)`)

And Line 104 and 105:
```python
hist_cen += np.histogramdd(data[cen_mask], bins)[0]
hist_sat += np.histogramdd(data[sat_mask], bins)[0]
```